### PR TITLE
[receiver/hostmetrics] root_path configuration and usage in filesystem scraper

### PR DESCRIPTION
**Pull request by jamesmoessis**
**Description:**

* Make the receiver configurable to be aware of the root filesystem, for Linux only.
* Give the ability to the filesystem scraper to translate mountpoints between host and container
  *  `mountpoint` attribute emitted without the root path prefix, i.e. from the host's perspective
  * No longer requires specification of where to find the `mountinfo` file through env var in order to avoid errors.
* All scrapers can see `RootPath` - useful for other scrapers e.g. `pagingscraper` as discussed in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5879
* Set the environment variables for the user if they are not set.
* Validate that the `root_path` aligns with the `gopsutil` env vars (if they are previously set). 
* Must keep track of a `globalRootPath` in order to ensure consistency between receivers and env vars. 

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5879

**Testing:**

Unit tests for all functionality. 

**Documentation:**

Documented how to collect host metrics from inside a container in the README.

### CODE REVIEW

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (dmitryax) [inline receiver/hostmetricsreceiver/hostmetrics_linux.go:26]: Should return if it's "" or "/"?

**Review by MovieStoreGuy — APPROVED** [review]
Documentation and changes make sense, just a comment on a type name but it is more of a nit than anything.

**Comment by MovieStoreGuy** [comment]
Don't forget to run `make chlog-new`

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (dmitryax) [inline receiver/hostmetricsreceiver/hostmetrics_linux.go:46]: Similar to my first comment, it's unclear why we require `HOST_PROC` to be set if rootPath is provided. I believe `rootPath` should always take precedence and set all the gopsutil env vars if they are empty.

File: receiver/hostmetricsreceiver/internal/scraper.go
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper.go:39]: Setter doesn't seem like a good name for a struct
```suggestion
type ScraperConfig struct {
```

File: receiver/hostmetricsreceiver/README.md
Comment (evan-bradley) [inline receiver/hostmetricsreceiver/README.md:147]: We should amend this to just say that the metrics are collected from system directories on the filesystem or something similar. Some of the scrapers in the host metrics receiver collect from directories other than `/proc`.

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go:88]: The way it currently works, it need both the translated and untranslated mountpoints. The current flow is: 

1. Get the mountpoints from `${root_path}/proc/1/mountinfo`
2. Those mountpoints, because they are delivered from the host, don't contain the `root_path` prefix, so that needs to be added, otherwise we would get a file not found error. Thus we get `translatedPartitions` which is prefixed with `root_path`.
3. When we want to record the metric, we want the `mountpoint` attribute to be from the host's perspective, not the container's perspective. So we use the original, untranslated mountpoint for that. 

There is the option of doing it the other way around too, i.e. get the mountpoints from the container proc and then remove the prefix when recording the metric. I would be interested to hear people's thoughts.

File: receiver/hostmetricsreceiver/README.md
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/README.md:172]: The problem then arises of ensuring that all instances of the host metrics receiver have the same `root_path`, which is kind of messy and requires global variables (an potentially a mutex, not sure if the instances are initialised async?). 

That being said, I do agree not having to worry about these variables is more ideal for the user. I'd be happy to add in setting of the environment variables, I just wanted to discuss it here first to make sure we are on the same page about it.

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/hostmetrics_linux.go:26]: I thought a better default for `RootPath` is `""`, because `/` is only the rootpath on unix environments. But I do agree, probably fine to return here if the user has explicitly set it as `/`. I will add it in.

File: receiver/hostmetricsreceiver/README.md
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/README.md:147]: Yeah, agreed. This could be more accurate. I will amend it.

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/hostmetrics_linux.go:45]: I think as discussed elsewhere, we will set the env vars from the receiver itself (if they aren't set). In that case if you wanted to ignore the feature, you just wouldn't set `root_path` in the config.

File: receiver/hostmetricsreceiver/internal/scraper.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/internal/scraper.go:39]: Think I'll go with `ParentConfig` as suggested by @dmitryax

File: receiver/hostmetricsreceiver/README.md
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/README.md:157]: Yes, I like that idea. Are you aware of which filesystems are used by which scrapers? From my quick look, I only saw calls to `/proc` + the filesystems that the user specifies in the filesystem scraper. Are there other ones?

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go:88]: I should also point out that if `root_path` is set, the user needs to specifies which filesystems to query from the host's perspective. This is because we expect the reported `mountpoint` attribute to be reported from the host's perspective. 

It's a little bit tricky, and there are a few ways it could be done. This is ultimately what I came to though.

File: receiver/hostmetricsreceiver/README.md
Comment (dmitryax) [reply receiver/hostmetricsreceiver/README.md:172]: > The problem then arises of ensuring that all instances of the host metrics receiver have the same root_path, which is kind of messy and requires global variables (an potentially a mutex, not sure if the instances are initialised async?).

We can give `root_path` config option higher precedence over the gopsutil env vars and set them from the config value if they're unset. In that case they can serve a role of that global variable / mutex. The question is if it's possible to run into a race condition with env var getters/setters. I believe that all the components are initialized sequentially in the same goroutine, so it should fine. In that case we need to make the error message a bit more generic and mention all possible conflicts. Let me know what do you think.

Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/README.md]: Ok, I've gone ahead and set the env variables, see my recent commits.

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/hostmetrics_linux.go]: Fixed it up, as mentioned elsewhere 👍

File: receiver/hostmetricsreceiver/internal/scraper.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/internal/scraper.go]: Done

File: receiver/hostmetricsreceiver/README.md
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/README.md]: Yeah, as mentioned elsewhere in this PR, we've gone with just `root_path` and are setting the env vars for the user. I've deleted this section of the doc.

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go
Comment (MovieStoreGuy) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go:88]: What you could do here instead of copying the mounts, is:

```golang
for i, parition := range parition {
  // Filtering here

  // Comment here about why the filepath.Join is needed here.
  usage, err := s.usage(filepath.Join(s.rootfs, partition.Mountpoint))
}
```
I suspect the above my not work for windows, but if the rootfs value is the zerovalue, the string should have no changes?

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (MovieStoreGuy) [inline receiver/hostmetricsreceiver/hostmetrics_linux.go:31]: `HOST_PROC_MOUNTINFO` is missing from this list, is that intention?

Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/hostmetrics_linux.go:31]: Yes that's intentional. The translations allow us to use the mountpoints under `HOST_PROC` so it doesn't need the alternate `mountinfo` file.

File: receiver/hostmetricsreceiver/README.md
Comment (dmitryax) [reply receiver/hostmetricsreceiver/README.md]: Looks great! Thanks

**Review by dmitryax — APPROVED** [review]
Looks good to me. Thank you for handling this! Waiting for approval from @MovieStoreGuy before merging

File: receiver/hostmetricsreceiver/README.md
Comment (MovieStoreGuy) [reply receiver/hostmetricsreceiver/README.md:157]: > Could we give more prescriptive advice around which host filesystems are used by which scrapers? I think the default recommendation should be to mount only the set of filesystems that gopsutil uses, and to optionally further restrict the mountpoints depending on the scrapers that are used. Mounting the entire root directory seems like it should be avoided in most cases.

I think this should be its own issue and marked for `good-first-issue` since this could be a lot of back and forth on what could be recommended.

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (MovieStoreGuy) [inline receiver/hostmetricsreceiver/hostmetrics_linux.go:37]: Is there a lock or atomic access on this value?

Just in case there several instances of it started at once?

**Review by MovieStoreGuy — COMMENTED** [review]
Just a comment on the global val to make sure it is safe for concurrent use, moving extended docs into its own issue because I think it would require some additional crowd sourcing.

File: receiver/hostmetricsreceiver/README.md
Comment (MovieStoreGuy) [reply receiver/hostmetricsreceiver/README.md:157]: @evan-bradley I've created the request here https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16093 to make sure this isn't lost.

File: receiver/hostmetricsreceiver/hostmetrics_linux.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/hostmetrics_linux.go:37]: I considered a lock, but decided it was unnecessary complexity. I checked the access and it is not accessed concurrently. `globalRootPath` is accessed solely from `Validate()` which is called sequentially for all configs from [this line of code](https://github.com/open-telemetry/opentelemetry-collector/blob/d6ed8246b64aba33d6b0f270f46a6b907b7f5fd5/service/config.go#L62) in the core collector.

**Comment by jamesmoessis** [comment]
Thanks @dmitryax for reviewing 😄 

I've addressed @MovieStoreGuy's comment. 

I also just pushed a small comment clarifying how users should configure the mount point filters in the filesystem scraper when `root_path` is set. 

I ran with the new config on a linux host and it's working well.

**Review by MovieStoreGuy — APPROVED** [review]
LGTM !

**Review by evan-bradley — APPROVED** [review]

File: receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go
Comment (jamesmoessis) [reply receiver/hostmetricsreceiver/internal/scraper/filesystemscraper/filesystem_scraper.go]: Talked with @MovieStoreGuy and implemented something similar to what he has here. Makes it cleaner!

**Review by dmitryax — APPROVED** [review]

**Pull request merged**
